goodtables/compat.py

```diff
@@ -30,22 +31,22 @@
     basestring = basestring
     numeric_types = (int, long, float)
 
-    def csv_reader(data, dialect=csv.excel, **kwargs):
+    def csv_reader(data, dialect=csv.excel, header_index=0, **kwargs):
         """Read text stream (unicode on Py2.7) as CSV."""
 
-        first_lines = list(islice(data, 5))
-        try:
-            dialect = csv.Sniffer().sniff(''.join(first_lines))
-            dialect.delimiter = dialect.delimiter.encode('utf-8')
-            dialect.quotechar = dialect.quotechar.encode('utf-8')
-        except csv.Error:
-            dialect = csv.excel
-
+        first_lines = list(islice(data, header_index, header_index + 2))
+        dialect = detect_dialect(first_lines)
+        
+        while not re.findall('[^\w ]', dialect.delimiter): 
+            first_lines.extend(list(data.readline()))
```

Comment:
because appending a list to a list creates a 2 dimensional list: 
```
In [7]: test_list = [2,3,4]
In [8]: test_list.append([6])
In [9]: test_list
Out[9]: [2, 3, 4, [6]]

In [10]: test_list.extend([8])
In [11]: test_list
Out[11]: [2, 3, 4, [6], 8]
```